### PR TITLE
[SPARK-8125] [SQL] Accelerates Parquet schema merging and partition discovery

### PR DESCRIPTION
**Pull request by liancheng**
This PR tries to accelerate Parquet schema discovery and `HadoopFsRelation` partition discovery.  The acceleration is done by the following means:
- Turning off schema merging by default
  
  Schema merging is not the most common case, but requires reading footers of all Parquet part-files and can be very slow.
- Avoiding `FileSystem.globStatus()` call when possible
  
  `FileSystem.globStatus()` may issue multiple synchronous RPC calls, and can be very slow (esp. on S3).  This PR adds `SparkHadoopUtil.globPathIfNecessary()`, which only issues RPC calls when the path contain glob-pattern specific character(s) (`{}[]*?\`).
  
  This is especially useful when converting a metastore Parquet table with lots of partitions, since Spark SQL adds all partition directories as the input paths, and currently we do a `globStatus` call on each input path sequentially.
- Listing leaf files in parallel when the number of input paths exceeds a threshold
  
  Listing leaf files is required by partition discovery.  Currently it is done on driver side, and can be slow when there are lots of (nested) directories, since each `FileSystem.listStatus()` call issues an RPC.  In this PR, we list leaf files in a BFS style, and resort to a Spark job once we found that the number of directories need to be listed exceed a threshold.
  
  The threshold is controlled by `SQLConf` option `spark.sql.sources.parallelPartitionDiscovery.threshold`, which defaults to 32.
- Discovering Parquet schema in parallel
  
  Currently, schema merging is also done on driver side, and needs to read footers of all part-files.  This PR uses a Spark job to do schema merging.  Together with task side metadata reading in Parquet 1.7.0, we never read any footers on driver side now.

### CODE REVIEW

**Comment by liancheng** [comment]
cc @marmbrus

**Comment by liancheng** [comment]
One of the test failure above is legitimate, which was caused by making `mergeSchema` to `false` by default, fixed. Others looked like to be affected by #7216.

**Comment by liancheng** [comment]
retest please

**Comment by liancheng** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala [hunk 1]
```diff
@@ -228,7 +228,7 @@ private[spark] object SQLConf {
     doc = "<TODO>")
 
   val PARQUET_SCHEMA_MERGING_ENABLED = booleanConf("spark.sql.parquet.mergeSchema",
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: Should this be a per relation option?

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: There is one. Defined in `object ParquetRelation2`, named `mergeSchema`.

**Comment by liancheng** [comment]
retest this please

**Comment by marmbrus** [comment]
Thanks, merging to master!

**Pull request closed without merging**